src/kernel.hpp

```diff
@@ -192,10 +193,13 @@ struct cvk_kernel_argument_values {
             if (size != sizeof(void*)) {
                 return CL_INVALID_ARG_SIZE;
             }
-
-            auto refc = *reinterpret_cast<refcounted* const*>(value);
-
-            m_kernel_resources[arg.binding].reset(refc);
+            if (arg.kind == kernel_argument_kind::sampler) {
+                m_kernel_resources[arg.binding].reset(static_cast<cvk_sampler*>(
+                    *reinterpret_cast<const cl_sampler*>(value)));
```

Comment:
I suggest doing this in two/three steps (value to api sampler, api sampler to sampler, use) for readability.